File: src/swift/3_StructuralPatterns/7_Proxy/ProtectionProxy.swift
import Foundation

protocol Vehicle
{
  func drive()
}

class Car : Vehicle
{
  func drive()
  {
    print("Car being driven")
  }
}

class CarProxy : Vehicle
{
  private let car = Car()
  private let driver: Driver

  init(driver: Driver)
  {
    self.driver = driver
  }

  func drive()
  {
    if driver.age >= 16
    {
      car.drive()
    }
    else {
      print("Driver too young")
    }
  }
}

class Driver
{
  var age: Int
  init(age: Int)
  {
    self.age = age
  }
}

func main()
{
  let car: Vehicle = CarProxy(driver: Driver(age: 12))
  car.drive()
}

main()

File: src/js/3_StructuralPatterns/7_Proxy/protection_proxy.js
class Car
{
  drive()
  {
    console.log('Car being driven');
  }
}

class CarProxy
{
  constructor(driver)
  {
    this.driver = driver;
    this._car = new Car();
  }

  drive()
  {
    if (this.driver.age >= 16)
      this._car.drive();
    else
      console.log('Driver too young');
  }
}

class Driver
{
  constructor(age)
  {
    this.age = age;
  }
}

let car = new Car();
car.drive();

let car2 = new CarProxy(new Driver(12)); // try 22
car2.drive();

